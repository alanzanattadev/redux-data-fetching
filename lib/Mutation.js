// @flow

export default class Mutation {
  mutationQL: string;
  variables: { [variableName: string]: mixed };
  onError: () => void;
  onCompleted: () => void;
  operationName: ?string;

  constructor(
    {
      mutationQL,
      variables = {},
      onError = () => {},
      onCompleted = () => {},
      operationName,
    }: {
      mutationQL: string,
      variables?: { [variableName: string]: mixed },
      onError?: () => void,
      onCompleted?: () => void,
      operationName?: string,
    } = {},
  ) {
    if (mutationQL == null) {
      throw new Error(
        "mutationQL has to be defined in the params of new Mutation: new Mutation({mutationQL: `mutation ...`})",
      );
    }
    this.mutationQL = mutationQL;
    this.variables = variables;
    this.onError = onError;
    this.onCompleted = onCompleted;
    this.operationName = operationName;
  }

  setOperationName(name: string) {
    return new Mutation({
      mutationQL: this.mutationQL,
      variables: this.variables,
      onError: this.onError,
      onCompleted: this.onCompleted,
      operationName: name,
    });
  }
}
