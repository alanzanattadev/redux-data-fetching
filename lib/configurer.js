"use babel";
// @flow
import configureMiddleware from "./middleware";
import configureReducer from "./reducer";
import configureActions from "./actions";
import { graphQLizr, graphQLRecordr } from "./graphqlTypesConverters";
import { schema } from "normalizr";
import { buildSchema } from "graphql";
import type { GraphQLSchema } from "graphql";

export function configure(
  graphQLSchema: GraphQLSchema | string,
  context: mixed,
) {
  const graphQLCompiledSchema = typeof graphQLSchema === "string"
    ? buildSchema(graphQLSchema)
    : graphQLSchema;
  const normalizrModel = graphQLizr(graphQLCompiledSchema);
  const recordsModel = graphQLRecordr(graphQLCompiledSchema);
  const actions = configureActions();
  return {
    actions,
    middleware: configureMiddleware(
      graphQLCompiledSchema,
      actions,
      normalizrModel,
      context,
    ),
    reducer: configureReducer(
      normalizrModel.entities,
      recordsModel,
      graphQLCompiledSchema,
    ),
    normalizrModel: normalizrModel,
    recordsModel: recordsModel,
  };
}
