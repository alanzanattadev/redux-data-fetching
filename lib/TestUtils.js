// @flow

import React from "react";
import { graphql } from "graphql";
import type { GraphQLSchema } from "graphql";
import Mutation from "./Mutation";
import { compose, withPropsOnChange, lifecycle } from "recompose";
import { DataReducerRecord } from "./reducer";

function configureProxies(
  {
    schema,
    rootValue,
    contextValue,
  }: {
    schema: GraphQLSchema,
    rootValue?: { [key: string]: (...params: any) => any },
    contextValue?: any,
  } = {},
) {
  return {
    configureDispatchProxy(
      {
        dispatch = () => {},
        onError = () => {},
        onSuccess = () => {},
        onResolve = () => {},
      }: {
        dispatch?: (action: Object) => void,
        onError?: (errors: Array<Object>) => void,
        onSuccess?: (data: Object) => void,
        onResolve?: (promise: Promise<any>) => void,
      } = {},
    ) {
      return function dispatchProxy(action) {
        if (action.type === "GRAPHQL_MUTATION") {
          const promise = graphql(
            schema,
            action.payload.mutation.mutationQL,
            rootValue,
            contextValue,
            action.payload.mutation.variables,
          );
          onResolve(promise);
          promise.then(result => {
            if (result.errors && result.errors.length > 0) {
              onError(result.errors);
            } else {
              dispatch(action);
              onSuccess(result.data);
            }
          });
        } else if (action.type === "GRAPHQL_FETCH") {
          const promise = graphql(
            schema,
            action.payload,
            rootValue,
            contextValue,
          );
          onResolve(promise);
          promise.then(result => {
            if (result.errors && result.errors.length > 0) {
              onError(result.errors);
            } else {
              dispatch(action);
              onSuccess(result.data);
            }
          });
        } else {
          dispatch(action);
        }
      };
    },
  };
}

const MockView = compose(
  lifecycle({
    componentDidMount() {
      if (typeof this.props.onMount === "function") {
        this.props.onMount(this.props);
      }
    },

    componentWillReceiveProps(nextProps) {
      if (typeof this.props.onPropChange === "function") {
        if (Array.isArray(this.props.watchedProps)) {
          const changed = this.props.watchedProps.reduce((red, prop) => {
            if (red === true) {
              return red;
            } else {
              return this.props[prop] !== nextProps[prop];
            }
          }, false);
          if (changed) {
            this.props.onPropChange(nextProps, this.props);
          }
        } else if (typeof this.props.watchedProps === "string") {
          if (
            nextProps[this.props.watchedProps] !==
            this.props[this.props.watchedProps]
          ) {
            this.props.onPropChange(nextProps, this.props);
          }
        } else {
          throw new Error(
            "No watch configured. Add watchedProps prop to your TestLab to determine which prop to observe for change",
          );
        }
      }
    },
  }),
)(function MockViewElement() {
  return null;
});

export function createTestLab(
  TestedHOC: (
    comp: Class<React.Component<any, any, any>>,
  ) => Class<React.Component<any, any, any>>,
  {
    reducerName = "data",
    schema,
  }: { reducerName?: string, schema: GraphQLSchema } = {},
): Class<React.Component<any, any, any>> {
  if (schema == null) {
    throw new Error(
      "You have to give your graphql schema to the lab. createTestLab(HOC, { schema: YOUR-SCHEMA })",
    );
  }
  const Subject = TestedHOC(MockView);
  const proxies = configureProxies({ schema });
  const DataLab = compose(
    withPropsOnChange(
      ["dispatch"],
      ({ dispatch, onResolve, onSuccess, onError }) => ({
        dispatch: proxies.configureDispatchProxy({
          dispatch,
          onResolve,
          onSuccess,
          onError,
        }),
        [reducerName]: new DataReducerRecord(),
      }),
    ),
  );
  const SupervisedSubject = compose(DataLab)(Subject);
  return SupervisedSubject;
}
