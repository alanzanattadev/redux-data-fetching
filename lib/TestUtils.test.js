// @flow

import React from "react";
import { mount } from "enzyme";
import { buildSchema } from "graphql";
import { createTestLab } from "./TestUtils";
import { configure } from "./configurer";
import Mutation from "./Mutation";

describe("TestUtils", () => {
  const schema = buildSchema(`
    type Query {
      users: [User]
    }
    
    type Mutation {
      createUser(name: String!, friends: [Int!]): User
    }
    
    type User {
      id: String!,
      name: String!,
      friends: [User!]
    }
  `);
  const rootValue = {
    users: () => [],
    createUser: ({ name, friends }) => ({
      id: "ah",
      name,
      friends: friends.map(friend => ({
        id: friend,
        name: "ok",
        friends: [],
      })),
    }),
  };
  const { DataHandlers, DataFetcher } = configure(schema, {}, rootValue);
  describe("createTestLab", () => {
    it("tests a correct DataHandlers correctly", done => {
      const Lab = createTestLab(
        DataHandlers({
          mapMutationsToProps: () => ({
            onSubmit: () =>
              new Mutation({
                mutationQL: `mutation {
              createUser(name: "Alan", friends: [2]) {
                id,
                name
              },
            }`,
              }),
          }),
        }),
        {
          schema,
          rootValue,
        },
      );
      const subject = mount(
        <Lab
          onMount={props => {
            props.onSubmit();
          }}
          onSuccess={done}
        />,
      );
    });

    it("tests a bad DataHandlers correctly", done => {
      const Lab = createTestLab(
        DataHandlers({
          mapMutationsToProps: () => ({
            onSubmit: () =>
              new Mutation({
                mutationQL: `mutation {
                  createUser(name: "Alan", friends: [2]),
                }`,
              }),
          }),
        }),
        {
          schema,
          rootValue,
        },
      );
      const subject = mount(
        <Lab
          onMount={props => {
            props.onSubmit();
          }}
          onError={errors => {
            done();
          }}
        />,
      );
    });

    it("tests a correct DataFetcher correctly", done => {
      const Lab = createTestLab(
        DataFetcher({
          mapPropsToNeeds: () => `{
          users {
            id,
            name,
            friends {
              id,
            }
          }
        }`,
        }),
        { schema, rootValue },
      );

      const subject = mount(
        <Lab
          watchedProps="users"
          onPropChange={props => {
            console.log(props.users);
            expect(props.users).toMatchSnapshot();
            done();
          }}
        />,
      );
    });

    it("tests a bad DataFetcher correctly", done => {
      const Lab = createTestLab(
        DataFetcher({
          mapPropsToNeeds: () => `{
          users(invalidArg: 1) {
            id,
            name,
            friends {
              id,
            }
          }
        }`,
        }),
        { schema, rootValue },
      );

      const subject = mount(
        <Lab
          watchedProps="users"
          onError={() => {
            done();
          }}
        />,
      );
    });
  });
});
