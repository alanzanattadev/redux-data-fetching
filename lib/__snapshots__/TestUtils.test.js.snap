// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestUtils createTestLab tests a correct DataFetcher correctly 1`] = `
Immutable.List [
]
`;
